lib/network_pos.go

```diff
@@ -0,0 +1,347 @@
+	// No versioning field is needed for this type since it is a member field
+	// for other top-level P2P messages, which will be versioned themselves.
+
+	// The block hash corresponding to the block that this QC authorizes.
+	BlockHash *BlockHash
+
+	// The view number when the the block was proposed.
+	ProposedInView uint64
+
+	// This BLS signature is aggregated from all of the partial BLS signatures for
+	// vote messages that have been aggregated by the leader. The partial signatures
+	// sign the (ProposedInView, BlockHash) pair for the block.
+	//
+	// From the block hash, we can look up the block height, the validator set at that
+	// block height, and the ordering of validators in consensus which identifies the
+	// present signers in the provided signers list. We can then use this to determine
+	// if the QC has 2/3rds of the total stake.
+	ValidatorsVoteAggregatedSignature *AggregatedBLSSignature
+}
+
+func (qc *QuorumCertificate) ToBytes() ([]byte, error) {
+	retBytes := []byte{}
+
+	// BlockHash
+	if qc.BlockHash == nil {
+		return nil, errors.New("QuorumCertificate.ToBytes: BlockHash must not be nil")
+	}
+	retBytes = append(retBytes, qc.BlockHash.ToBytes()...)
+
+	// ProposedInView
+	retBytes = append(retBytes, UintToBuf(qc.ProposedInView)...)
+
+	// ValidatorsVoteAggregatedSignature
+	if qc.ValidatorsVoteAggregatedSignature == nil {
+		return nil, errors.New("QuorumCertificate.ToBytes: ValidatorsVoteAggregatedSignature must not be nil")
+	}
+	encodedValidatorsVoteAggregatedSignature, err := qc.ValidatorsVoteAggregatedSignature.ToBytes()
+	if err != nil {
+		return nil, errors.Wrapf(err, "QuorumCertificate.ToBytes: Error encoding ValidatorsVoteAggregatedSignature")
+	}
+	retBytes = append(retBytes, encodedValidatorsVoteAggregatedSignature...)
+
+	return retBytes, nil
+}
+
+func DecodeQuorumCertificate(rr io.Reader) (*QuorumCertificate, error) {
+	var qc QuorumCertificate
+	var err error
+
+	qc.BlockHash, err = ReadBlockHash(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DecodeQuorumCertificate: Error decoding BlockHash")
+	}
+
+	qc.ProposedInView, err = ReadUvarint(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DecodeQuorumCertificate: Error decoding ProposedInView")
+	}
+
+	qc.ValidatorsVoteAggregatedSignature, err = DecodeAggregatedBLSSignature(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DecodeQuorumCertificate: Error decoding ValidatorsVoteAggregatedSignature")
+	}
+
+	return &qc, nil
+}
+
+// This is an aggregated BLS signature from a set of validators. Each validator's
+// presence in the signature is denoted in the provided signers list. I.e. if the
+// list's value at index 0 is 1, then the validator identified by that index is
+// present in the aggregated signature. The indices of all validators are expected
+// to be known by the caller.
+type AggregatedBLSSignature struct {
+	// TODO: Switch this to a bitlist, which will result in ~8x reduction in total
+	// size of this construct.
```

Comment:
I came across this BitSet package when coding the EF encoding. It had everything I needed:
https://pkg.go.dev/github.com/bits-and-blooms/bitset

Reply:
SG. I'll switch to a bitlist in a separate PR.

Reply:
Implemented in https://github.com/deso-protocol/core/pull/564